### PR TITLE
Implement PayableAddressStore for Ledger Accounts

### PR DESCRIPTION
**Pull request by topocount**
This is the first in a series of PRs that will implement single-click reward distributions for maintainers.

Note that this is functionally distinct from the ethereum plugin. That plugin is meant for users to authenticate via an EOA (externally owned account). The `PayableAddressStore` is tightly coupled to the ledger and allows users to configure the accounts they want grain paid out to.

Future work includes:
-  creating the config for the distribution and/or buyback currencies
- writing the dapp code that is capable of distributing grain payout and buybacks onchain
- writing the EVM event parsing code that verifies the distribution or buyback, and then writes the
appropriate activity to the Ledger.

PayableAddressStore maps addresses of EVM-based currencies to a participant's
address capable of accepting the currency. This structure exists to
accomodate safe migration for grain/payout token changes. Users must verify
themselves that the address they are supplying is capable of receiving their
share of a grain distribution.

An address may be deleted by passing in an empty string for the
`address` parameter. This is useful in case the underlying private key
is compromised or the exchange hosting a custodial account is hacked.

Merging Identities has no effect on the base Identity's configured payable addresses. The target
identity's address are dropped and all payouts after this point will target the base account's configured
address.

The CurrencyId objects are currently serialized before use in the Map.
This is due to the fact that Objects used as keys must be directly
accessible in order to retrieve the mapped value. A new object with
matching members is not equal to the object used as the key because each
has a distinct memory reference. By stringifying the object before
entering it as the key, deep equality can be observed and the
composition of the object become the keys, rather than the object's
memory reference.

test plan: Unit test are provided

### CODE REVIEW

File: src/core/ledger/ledger.js
Comment (decentralion) [inline src/core/ledger/ledger.js:619]: why pass in `""` as the delete param instead of `null`?

File: src/core/ledger/ledger.js
Comment (decentralion) [inline src/core/ledger/ledger.js:647]: having a method throw an error after a mutation is scary in this module, since it may leave the ledger in an inconsistent state. please follow the convention from the other methods, and have a clear demarcation between the argument-checking logic and error handling (above the demarcation line) and any mutation (below the demarcation line).

(in this case, since the mutation that is above the error checking returns immediately, it's safe, but it's not obviously safe, and it's easy to imagine someone in the future making an innocent-looking change that actually makes the code unsafe.)

File: src/core/ledger/ledger.test.js
Comment (decentralion) [inline src/core/ledger/ledger.test.js:807]: we don't re-use mutable state between test cases. it invites "spooky error at a distance" behavior where one test could pass while breaking a subsequent test. please refactor this so that each test has its own independent ledger instance, see other tests for examples.

**Review by decentralion — COMMENTED** [review]
this solution seems to have good support for tracking multiple ERC-20s, which is nice. but it's super coupled to the EVM. what if we wanted to use it to represent FIL or BTC?

i'm wondering if we could have something as simple as having currencyAddress be an object that has a "chain identifier" (e.g. BTC, ETH, XDAI) and a "currency identifier". When the chain identifier points to "ETH" or "XDAI", then the currency identifier points to the right contract. If the chain identifier is BTC, then the currency identifier can be superfluous.

File: src/core/ledger/ledger.js
Comment (topocount) [reply src/core/ledger/ledger.js:619]: type-consistency but null makes sense

**Comment by decentralion** [comment]
Removing myself from the review -- I think @hammadj is a better reviewer since he's more invested in this feature path

File: src/core/ledger/currency.js
Comment (hammadj) [inline src/core/ledger/currency.js:32]: Maybe `evmToken` would be a better name?

File: src/core/ledger/currency.js
Comment (hammadj) [inline src/core/ledger/currency.js:63]: If its unused, why have it declared?

File: src/core/ledger/currency.js
Comment (hammadj) [inline src/core/ledger/currency.js:76]: ProtocolSymbol being assigned to chainId is a bit confusing. Feel like it might be cleaner to have a single type for all tokens / currencies, something like:

```ts
declare type Currency = {
  symbol: string,
  chainId?: EvmChainId,
  tokenAddress?: EvmAddress
}
```

Where `chainId` and `tokenAddress` would be undefined for non-EVM currencies.

File: src/core/ledger/ledger.js
Comment (hammadj) [inline src/core/ledger/ledger.js:613]: you mention an `address` param but there isn't one, should be `payoutAddress`. Also, instead of an empty string should it be passing null since that's a valid type for that param?

File: src/core/ledger/ledger.js
Comment (hammadj) [inline src/core/ledger/ledger.js:627]: Would be cleaner to take in the `Currency` object type I described above instead, something like this:
```ts
  setPayoutAddress(
     id: IdentityId,
     currency: Currency,
     payoutAddress: PayoutAddress | null,
  ): Ledger {
    this._createAndProcessEvent({
      type: "SET_PAYOUT_ADDRESS",
      id,
      currency,
      payoutAddress,
    });
```

Also, this structure implies that different people would be able to get payouts in different tokens on different chains. Is there any use case where you think that might be relevant? Feels redundant to have the currency object in every ledger entry for setting a payout address.

File: src/core/ledger/currency.js
Comment (hammadj) [inline src/core/ledger/currency.js:79]: Likewise instead of `currencyIdParser` , it can just be `currencyParser`

File: src/core/ledger/ledger.js
Comment (hammadj) [inline src/core/ledger/ledger.js:52]: Could be confused for nodeAddresses or something else. `payoutAddresses` would be a better name

File: src/core/ledger/ledger.test.js
Comment (hammadj) [inline src/core/ledger/ledger.test.js:852]: These two tests are skipped, any reason why or just forgot to un-skip?

File: src/core/ledger/ledger.test.js
Comment (hammadj) [inline src/core/ledger/ledger.test.js:794]: Should also add a tests to ensure payoutAddresses are preserved after merging identities:
- identityA with address merged into identityB without address
- identityA without address merged into identityB with address
- identityA with address merged into identityB with address

**Review by hammadj — COMMENTED** [review]
few naming / test things, overall looks good! this will help a ton

File: src/core/ledger/currency.js
Comment (topocount) [reply src/core/ledger/currency.js:76]: @hammadj I considered this monolithic type but I thought is was more confusing for static typing. By splitting between these two subtypes, it's clearer in flow what type we're expecting and what sort of details it should possess.

File: src/core/ledger/currency.js
Comment (topocount) [reply src/core/ledger/currency.js:43]: EthAddress is "more" correct since the base currency isn't technically a token

File: src/core/ledger/currency.js
Comment (topocount) [reply src/core/ledger/currency.js:32]: Same explanation as above.

File: src/core/ledger/currency.js
Comment (topocount) [reply src/core/ledger/currency.js:63]: This was removed but I never rebased this code

File: src/core/ledger/ledger.js
Comment (topocount) [reply src/core/ledger/ledger.js]: @hammadj I considered this but I feel like this is leaking implementation details. Consumers of this API to only need to pass in `ChainId`s and and and an optional ethereum address to generate a `CurrencyId` internally.

File: src/core/ledger/currency.js
Comment (hammadj) [reply src/core/ledger/currency.js]: The vast majority of SC users would not be distributing Ether, they would be distributing some sort of ERC-20 token. Also, I was referring to the `currencyAddress` field and not `EthAddress` type.

File: src/core/ledger/currency.js
Comment (hammadj) [reply src/core/ledger/currency.js]: Not sure I follow. `EvmId` implies that this type is an ID of some sort when in reality it's an object that contains information about some token and the chain it's running on. The EvmId name might be more appropriate if this type was a string in the form of `"evm:chainId:tokenAddress"` and coming out of a parser, but since its an object calling it an ID has a smell to it

File: src/core/ledger/currency.js
Comment (hammadj) [reply src/core/ledger/currency.js]: Instead of doing the Union at the ID level, could just do something like this:

```ts

type EvmToken = {
  symbol: string,
  chainId: EvmChainId,
  tokenAddress: EvmAddress
}

type ProtocolToken = {
  symbol: string,
}

type Currency = EvmToken | ProtocolToken
```

This would also allow us to add additional fields scoped to different protocol tokens (e.g. if Polkadot has something like ERC20's on their network then we could have a tokenAddress field on ProtocolToken as well but enforce it to follow the semantics of its protocol instead of allowing a EvmChainId to be paired with a ProtocolId type)

File: src/core/ledger/ledger.js
Comment (hammadj) [reply src/core/ledger/ledger.js]: Wouldn't the `symbol` be a valuable thing to have in there as well, especially for rendering user-readable ledger history and other UI elements?

File: src/core/ledger/currency.js
Comment (hammadj) [inline src/core/ledger/currency.js:89]: Calling this CurrencyId makes me thing it behaves like a string since almost all instances of "IDs" that I can think of have been strings

File: src/core/ledger/ledger.js
Comment (hammadj) [inline src/core/ledger/ledger.js:88]: It would be a lot more intuitive id `CurrencyId` was the string that's generated from a `Currency` object

**Review by hammadj — COMMENTED** [review]
Still think the naming needs some adjustment but happy to reconsider if other devs disagree. I know this PRs been open a while so don't want to be too nitty, but I think naming is super important to get right / consistent. If you feel strongly about your naming and no one else feels strongly about my suggestions, then I'm happy to yield to your discretion :)

File: src/core/ledger/ledger.js
Comment (hammadj) [inline src/core/ledger/ledger.js:985]:
```suggestion
  +currency: Currency,
```

File: src/core/ledger/ledger.js
Comment (hammadj) [inline src/core/ledger/ledger.js:992]: same here, should be `currency`

File: src/core/ledger/currency.js
Comment (hammadj) [inline src/core/ledger/currency.js:72]: Much nicer :) Laaaaast nit would be to name this field `tokenAddress` or `tokenContract` instead of `currencyAddress` since that's the standard name used for this field in most web3 project APIs / docs etc.

e.g. from OpenSea JS: 
image

From Balancer:
image

From Uniswap:
image

**Review by hammadj — APPROVED** [review]
Approving with one last nit, feel free to merge after addressing

File: src/core/ledger/currency.js
Comment (topocount) [reply src/core/ledger/currency.js:72]: @hammadj I'd be happy to use `token` and `tokenAddress` but we have a special case for native ETH itself. (or xDAI on the xDai sidechain) and I want to be clear that we're accounting for those. We also have other special cases like SAI which don't conform to the ERC-20 standard and therefore begs the question: Is SAI actually a token if it's not ERC-20 conformant?

**Pull request merged**
